### PR TITLE
feat: Make garbage collection aware of cold storage

### PR DESCRIPTION
**Pull request by wacban**
Looking for early feedback while I'm figuring out how to write a nice unit test for it. 

Would you, dear reviewer, mind if I remove the cold_store rust feature altogether? It's annoying and in this PR would cause a circular dependency. The cold store implementation can still be gated by the config. 

There are three important changes in this PR:
- When updating the cold head update it in both hot and cold stores. 
- Enabling full garbage collection on archival nodes that are fully migrated to split storage. 
- Adjusting the garbage collection stop height to take cold store head into account.

### CODE REVIEW

**Comment by wacban** [comment]
I'm having second thoughts about removing the cold_store feature here:
- I managed to make this PR work without removing it
- Removing it is non-trivial and would require some thought and effort while I want to merge this pr soon to unblock other work.

I still think we should get rid of it sooner than later and just rely on config to gate the cold storage. The main benefits are:
- The cold store code would be exercised under all of the tests that we have.
- It would be more convenient to work with cold store code. 
- I think small incremental but fully tested changes are better than just releasing the whole thing all at once.

**Comment by nikurt** [comment]
> Removing the `cold_store` feature

Sounds good to me. The rust feature is needed for protocol changes, but for new functionality controlled by a config, that should be fine.

File: core/store/src/trie/mod.rs [hunk 1]
```diff
@@ -322,8 +322,11 @@ impl std::fmt::Debug for TrieNode {
 #[derive(Debug, Eq, PartialEq)]
 #[allow(clippy::large_enum_variant)]
 pub enum RawTrieNode {
+    /// Leaf(key, value_length, value_hash)
```
Comment (nikurt) [inline core/store/src/trie/mod.rs:325]: Thank you!

File: integration-tests/src/tests/client/cold_storage.rs [hunk 2]
```diff
@@ -201,24 +202,26 @@ fn test_cold_db_head_update() {
...
-        );
+        let head = &env.clients[0]
+            .runtime_adapter
+            .store()
+            .get_ser::<Tip>(DBCol::BlockMisc, HEAD_KEY)
+            .unwrap();
+        let cold_head_in_hot = hot_store.get_ser::<Tip>(DBCol::BlockMisc, COLD_HEAD_KEY).unwrap();
+        let cold_head_in_cold = cold_store.get_ser::<Tip>(DBCol::BlockMisc, HEAD_KEY).unwrap();
+
+        assert_eq!(head, &cold_head_in_cold, "checking cold_head_in_cold");
```
Comment (nikurt) [inline integration-tests/src/tests/client/cold_storage.rs:224]: The assert error message is tautological. The assertion by itself reports that `cold_head_in_cold` is different from `head`. The error message doesn't add anything new.

**Review by nikurt — COMMENTED** [review]
Looks good, waiting for a unit test, if possible.

**Review by nikurt — APPROVED** [review]

**Pull request merged**